torch/csrc/jit/passes/annotate_effects.cpp

```diff
@@ -0,0 +1,184 @@
+#include "torch/csrc/jit/passes/annotate_effects.h"
+#include <set>
+
+namespace torch {
+namespace jit {
+namespace {
+
+// Ops with effects that we need to track through the passing of world tokens
+// e.g. mutation
+const std::unordered_set<Symbol> effectfulOps = {
+    aten::append,
+};
+
+/**
+ * AnnotateEffects
+ *
+ * This pass annotates effectful operations (such as ones that mutate existing
+ * values) to prevent subsequent passes from re-ordering ops in a way that
+ * changes the meaning of the program.
+ *
+ * It does this by adding primitive memory fence ops around nodes that use
+ * mutable values. The memory fences input and output a "World" token, which
+ * expresses effects explicitly in the IR and forces all fenced nodes to be
+ * linearized.
+ */
+class AnnotateEffectsImpl {
+ public:
+  void annotateEffects(Graph* g) {
+    // Generate the first world token
+    const auto tokenGenerator = g->create(prim::Constant);
```

Comment:
I'm pretty sure this will work badly with inlining of function calls. The token should be passed from the caller to the callee, instead of callee having its own independent copy.

Reply:
yeah, I wasn't sure whether we wanted to require that side effects be contained within a single function. Will change